test/mempool-test.js

```diff
@@ -1055,4 +1056,175 @@ describe('Mempool', function() {
+
+    after(async () => {
+      await workers.close();
+      await chain.close();
+      await mempool.close();
+      await blocks.close();
+    });
+
+    // Number of coins available in
+    // chaincoins (100k satoshi per coin).
+    const N = 100;
+    const chaincoins = new MemWallet();
+    const wallet = new MemWallet();
+
+    it('should create txs in chain', async () => {
+      const mtx = new MTX();
+      mtx.addInput(new Input());
+
+      for (let i = 0; i < N; i++) {
+        const addr = chaincoins.createReceive().getAddress();
+        mtx.addOutput(addr, 100000);
+      }
+
+      const cb = mtx.toTX();
+      const block = await getMockBlock(chain, [cb], false);
+      const entry = await chain.add(block, VERIFY_NONE);
+
+      await mempool._addBlock(entry, block.txs);
+
+      // Add 100 blocks so we don't get premature
+      // spend of coinbase.
+      for (let i = 0; i < 100; i++) {
+        const block = await getMockBlock(chain);
+        const entry = await chain.add(block, VERIFY_NONE);
+
+        await mempool._addBlock(entry, block.txs);
+      }
+
+      chaincoins.addTX(cb);
+    });
+
+    it('should limit mempool size', async () => {
+      let expectedSize = 0;
+
+      for (let i = 0; i < N; i++) {
+        // Spend a different coin each time to avoid exceeding max ancestors.
+        const coin = chaincoins.getCoins()[i];
+        const addr = wallet.createReceive().getAddress();
+
+        const mtx = new MTX();
+        mtx.addCoin(coin);
+        // Increment fee with each TX so oldest TX gets evicted first.
+        // Otherwise the new TX might be the one that gets evicted,
+        // resulting in a "mempool full" error instead.
+        mtx.addOutput(addr, 90000 - (10 * i));
+        chaincoins.sign(mtx);
+        const tx = mtx.toTX();
+
+        expectedSize += txMemUsage;
+
+        if (expectedSize < maxSize) {
+          await mempool.addTX(tx);
+        } else {
+          assert(i >= 9);
+          let evicted = false;
+          mempool.once('remove entry', () => {
+            evicted = true;
+            // We've exceeded the max size by 1 TX
+            // Mempool will remove 2 TXs to get below 90% limit.
+            expectedSize -= txMemUsage * 2;
+          });
+          await mempool.addTX(tx);
+          assert(evicted);
+        }
+      }
+    });
+
+    it('should evict old transactions', async () => {
+      // Clear mempool. Note that TXs in last test were not
+      // added to the wallet: we can re-spend those coins.
+      await mempool.reset();
+
+      let now = 0;
+      const original = util.now;
+      try {
+        util.now = () => {
+          return now;
+        };
+
+        // After we cross the expiry threshold, one TX at a time
+        // will start to expire, starting with the oldest.
+        const sent = [];
+        mempool.on('remove entry', (entry) => {
+          const expected = sent.shift();
+          assert.bufferEqual(entry.tx.hash(), expected);
```

Comment:
need to add a counter here to assert later that the event actually fired at all.